### PR TITLE
[ENH] Widget Insertion

### PR DESCRIPTION
**Pull request by irgolic**
##### Feature
Insert a widget in-between two already connected widgets by either:
- dragging it from the widget selection menu onto the link,
- right-clicking the link and selecting 'Insert Widget'

##### Description of changes
- Added an 'Insert Node' command for purposes of undo/redo
- Added entry to 'right-click on link' context menu
- Widget drag from menu -> insert

##### Includes
- [X] Code changes
- [ ] Tests
- [X] Documentation

### CODE REVIEW

File: Orange/canvas/document/commands.py [hunk 1]
```diff
@@ -69,6 +71,38 @@ def undo(self):
         self.scheme.add_link(self.link)
 
 
+class InsertNodeCommand(QUndoCommand):
+    def __init__(self, scheme, link, new_node, parent=None):
+        QUndoCommand.__init__(self, "Remove link", parent)
```
Comment (ales-erjavec) [inline Orange/canvas/document/commands.py]: Change the command name/text to 'Insert widget on link' or something similar.

File: Orange/canvas/document/schemeedit.py [hunk 2]
```diff
@@ -1029,6 +1047,30 @@ def changeEvent(self, event):
 
         QWidget.changeEvent(self, event)
 
+    def tryInsertNode(self, link, new_node_desc, pos):
+        source_node = link.source_node
+        sink_node = link.sink_node
+
+        if nodes_are_compatible(source_node.description, new_node_desc) and \
+                nodes_are_compatible(new_node_desc, sink_node.description):
```
Comment (ales-erjavec) [inline Orange/canvas/document/schemeedit.py]: The test here should also be constrained on the existing `link.source_channel` and `link.sink_channel`.

E.g. If I use:

screen shot 2018-08-02 at 11 21 41

and drop an *Impute* widget on the link, I get a
```
------------------- IncompatibleChannelTypeError Exception --------------------
Traceback (most recent call last):
  File "/Users/aleserjavec/workspace/orange3/Orange/canvas/document/schemeedit.py", line 1102, in eventFilter
    self.tryInsertNode(link, desc, pos)
  File "/Users/aleserjavec/workspace/orange3/Orange/canvas/document/schemeedit.py", line 1066, in tryInsertNode
    new_node, possible_links[0][0][1]),  # first link, first entry, output
  File "/Users/aleserjavec/workspace/orange3/Orange/canvas/scheme/link.py", line 125, in __init__
    % (source_channel.type, sink_channel.type)
Orange.canvas.scheme.errors.IncompatibleChannelTypeError: Cannot connect 'Orange.base.Model' to 'Orange.base.Learner'
-------------------------------------------------------------------------------
```
The set of possible links below should also be filtered correspondingly.

P.S.
The same also applies to the 'Insert widget` context menu action.

**Comment by ales-erjavec** [comment]
Can you also implement drop target detection in `eventFilter` for `GraphicsSceneDragEnter`, `GraphicsSceneDragMove` and `GraphicsSceneDragLeave` events and highlight the drop target.
Something like:
```diff
diff --git a/Orange/canvas/document/schemeedit.py b/Orange/canvas/document/schemeedit.py
index 840ece304..ddf453fa6 100644
--- a/Orange/canvas/document/schemeedit.py
+++ b/Orange/canvas/document/schemeedit.py
@@ -132,6 +132,7 @@ class SchemeEditWidget(QWidget):
         self.__possibleMouseItemsMove = False
         self.__itemsMoving = {}
         self.__contextMenuTarget = None
+        self.__dropTarget = None
         self.__quickMenu = None
         self.__quickTip = ""
 
@@ -1066,18 +1067,43 @@ class SchemeEditWidget(QWidget):
 
     def eventFilter(self, obj, event):
         # Filter the scene's drag/drop events.
+        MIME_TYPE = "application/vnv.orange-canvas.registry.qualified-name"
         if obj is self.scene():
             etype = event.type()
             if etype == QEvent.GraphicsSceneDragEnter or \
                     etype == QEvent.GraphicsSceneDragMove:
                 mime_data = event.mimeData()
-                if mime_data.hasFormat(
-                        "application/vnv.orange-canvas.registry.qualified-name"
-                        ):
+                droptarget = None
+                accept = True
+                if mime_data.hasFormat(MIME_TYPE):
+                    qname = bytes(mime_data.data(MIME_TYPE)).decode("ascii")
+                    try:
+                        desc = self.__registry.widget(qname)
+                    except KeyError:
+                        pass
+                    else:
+                        item = self.__scene.item_at(event.scenePos(),
+                                                    items.LinkItem)
+                        link = self.__scene.link_for_item(item) if item else None
+                        if link is not None:
+                            droptarget = item
+                            droptarget.setHoverState(True)
+                            accept = can_insert_node(desc, link)
+                if accept:
                     event.acceptProposedAction()
                 else:
                     event.ignore()
+                if self.__dropTarget is not None and \
+                        self.__dropTarget is not droptarget:
+                    self.__dropTarget.setHoverState(False)
+                    self.__dropTarget = None
+
+                self.__dropTarget = droptarget
                 return True
+            elif etype == QEvent.GraphicsSceneDragLeave:
+                if self.__dropTarget is not None:
+                    self.__dropTarget.setHoverState(False)
+                    self.__dropTarget = None
```

**Comment by ajdapretnar** [comment]
Dragging the widget no longer works for me...?

**Comment by irgolic** [comment]
Are you dragging a compatible widget to the link? If the widget is incompatible, it's simply added as if dragged to empty canvas.
https://gyazo.com/bcd54f4c9279e804ed43f3629e938530

**Comment by ajdapretnar** [comment]
Yes. File - Select Columns - Data Table.

**Comment by irgolic** [comment]
Resolved.

**Pull request merged**